activemodel/lib/active_model/secure_password.rb

```diff
@@ -132,7 +132,21 @@ def has_secure_password(attribute = :password, validations: true)
             end
           end
 
-          validates_length_of attribute, maximum: ActiveModel::SecurePassword::MAX_PASSWORD_LENGTH_ALLOWED
+          # Validates that the password does not exceed the maximum allowed characters (72 characters) and
+          # the maximum allowed bytes (72 bytes) for BCrypt. The character length validation is checked first
+          # to provide a more user-friendly error message. However, the byte size validation is still necessary
+          # due to BCrypt's inherent limitation of 72 bytes.
+          validate do |record|
+            password_value = record.public_send(attribute)
+            if password_value.present?
+              if password_value.length > ActiveModel::SecurePassword::MAX_PASSWORD_LENGTH_ALLOWED
```

Comment:
Actually, this message is not user friendly. How many people know what a byte means exactly? I think we should just make the message more generic and cover both cases.

Reply:
For me, and what I understand, we now get two different errors:

* max by size
* max length

If my code generate a lenght ok, but a byte size over limit, I probably will look to the lenght, that is normal way to debug, and is the legacy behaviour...

Now, how that we are adding a new error validation, is interesting for me to add a new error message...

Probably in a debug, the developer will seed that this error have this new validation too, and will not have to spend more time debugging it... and will help in a future framework update debugging too

Reply:
@rafaelfranca that's right, the second message isn't friendly, that's why the first conditional exists just to keep the message "actionable" (you have a better idea about what to do to get rid of the error) for most of the cases.

We can have just one generic message (e.g "is too long, reduce number of characters") but it will be a "less-clear" message for most of the users.

Reply:
Urgh, now that I see that we can have cases where the password is 25 chars long by 73 bytes, making a generic message seems harder. The fix could be shortening the length of the password by 1 characters, or by 2, or by 3, or by the less than the half of it.

I do think we should just check the bytesize and ask people to shorten the password instead of saying what exactly is the limit.

Reply:
My take is that telling people "exceeded max 73 bytes" is fine, if you give them perspective.

ie. "exceeded limit of 73 bytes by 2 bytes (75 bytes)."

Also, if we apply this to the max, shouldn't there also be a minimum limit? 🤔

Reply:
@rafaelfranca I'll keep the generic message as "is too long"

@zzak the problem with showing "exceeded limit of 73 bytes by 2 bytes (75 bytes)" is that it is still unfriendly and incomprehensible for most of the users and doesn't give a better idea about how to fix the problem (e.g will leave they wondering about what they need to do to reduce it 2 bytes), I think going for the generic message is the best compromise we can reach on this.

In the future we might want to consider switching to alternatives algorithms without this limitation.